src/toxcore/dht_new/kbucket.rs

```diff
@@ -0,0 +1,857 @@
+                // index is pointing past the end
+                if self.is_full() {
+                    debug!(target: "Bucket",
+                        "Node is too distant to add to the bucket.");
+                    false
+                } else {
+                    // distance to the PK was bigger than the other keys, but
+                    // there's still free space in the bucket for a node
+                    debug!(target: "Bucket",
+                        "Node inserted at the end of the bucket.");
+                    self.nodes.push(*new_node);
+                    true
+                }
+            },
+            Err(index) => {
+                // index is pointing inside the list
+                if self.is_full() {
+                    debug!(target: "Bucket",
+                        "No free space left in the bucket, the last node removed.");
+                    self.nodes.pop();
+                }
+                debug!(target: "Bucket", "Node inserted inside the bucket.");
+                self.nodes.insert(index, *new_node);
+                true
+            },
+        }
+    }
+
+    /** Remove [`PackedNode`](./struct.PackedNode.html) with given PK from the
+    `Bucket`.
+
+    Note that you must pass the same `base_pk` each call or the internal
+    state will be undefined. Also `base_pk` must be equal to `base_pk` you added
+    a node with. Normally you don't call this function on your own but Kbucket does.
+
+    If there's no `PackedNode` with given PK, nothing is being done.
+    */
+    pub fn remove(&mut self, base_pk: &PublicKey, node_pk: &PublicKey) {
+        trace!(target: "Bucket", "Removing PackedNode with PK: {:?}", node_pk);
+        match self.nodes.binary_search_by(|n| base_pk.distance(n.pk(), node_pk) ) {
+            Ok(index) => {
+                self.nodes.remove(index);
+            },
+            Err(_) => {
+                trace!("No PackedNode to remove with PK: {:?}", node_pk);
+            }
+        }
+    }
+
+    /// Check if node with given PK is in the `Bucket`.
+    pub fn contains(&self, pk: &PublicKey) -> bool {
+        self.iter().any(|n| n.pk() == pk)
+    }
+
+    /// Get the capacity of the Bucket.
+    pub fn capacity(&self) -> usize {
+        self.capacity as usize
+    }
+
+    /** Check if `Bucket` is empty.
+
+    Returns `true` if there are no nodes in the `Bucket`, `false`
+    otherwise.
+    */
+    pub fn is_empty(&self) -> bool {
+        self.nodes.is_empty()
+    }
+
+    /** Check if `Bucket` is full.
+
+    Returns `true` if there is no free space in the `Bucket`, `false`
+    otherwise.
+    */
+    pub fn is_full(&self) -> bool {
+        self.nodes.len() == self.capacity()
+    }
+
+    /// Returns an iterator over contained [`PackedNode`]
+    /// (./struct.PackedNode.html)s.
+    pub fn iter(&self) -> BucketIter {
+        BucketIter { iter: self.nodes.iter() }
+    }
+}
+
+/// Iterator over `Bucket`.
+pub struct BucketIter<'a> {
```

Comment:
Do we really need BucketIter?

Reply:
It is used in https://github.com/NamsooCho/tox/blob/d29e7f1e9e7bccb148cad72b5a628b2c91f32414/src/toxcore/dht_new/kbucket.rs#L249

Reply:
https://github.com/NamsooCho/tox/blob/d29e7f1e9e7bccb148cad72b5a628b2c91f32414/src/toxcore/dht_new/kbucket.rs#L277-L279

Reply:
So we may replace it with `self.nodes.iter()` ?